[Task-7] As a user, I can get the reset password instruction via email

gutakk opened the pull request:
Resolve #8 

## What happened 👀

- Add forgot password page and related components without styling
- Add forgot password functionalities
- Fix alert component to be able to render different icons.
- Add forgot password as a public route
- Back will redirect to home page
- Add unit tests
 
## Insight 📝
N/A
 
## Proof Of Work 📹
- Send the recovery instruction email
![Screen Shot 2564-06-21 at 13 32 06](https://user-images.githubusercontent.com/29707647/122716916-15a49600-d295-11eb-83ca-f95ef9ffff93.png)

- Email
![Screen Shot 2564-06-21 at 13 32 58](https://user-images.githubusercontent.com/29707647/122717004-353bbe80-d295-11eb-9956-e6d7393e09c8.png)

- Validation
![Screen Shot 2564-06-21 at 13 33 12](https://user-images.githubusercontent.com/29707647/122717029-3cfb6300-d295-11eb-9918-61e1708cf7a0.png)

Commit:
[Task-7] Add forgot password component

Commit:
[Task-7] Add AuthInputField

Commit:
[Task-7] Add forgot password form

Commit:
[Task-7] Update sign in to use auth input field

Commit:
[Task-7] Add forgot password handler and adapter

Commit:
[Task-7] Add forgot password adapter and handler

Commit:
[Task-7] Add back button

Commit:
[Task-7] Add back function

Commit:
[Task-7] Remove key value return

Commit:
[Task-7] Add unit test for AuthInputField

Commit:
[Task-7] Add back icon and unit test for BackButton

Commit:
[Task-7] Get from test id instead

Commit:
[Task-7] Add TODO

Commit:
[Task-7] Add forgot password page test

Commit:
[Task-7] Add forgot password form unit test

Commit:
[Task-7] Fix linting

Commit:
[Task-7] Update data-test-id to data-testid

Commit:
[Task-7] Update data-testid to data-test-id and add config to let react testing library look for data-test-id

Commit:
[Task-7] Add alert unit test

Commit:
[Task-7] Add break line

Commit:
[Task-7] Update back button to redirect to home page

Commit:
[Task-7] Fix lint

rafayet-monon reviewed (commented):
Please add appropriate PR descriptions.

gutakk commented:
> Please add appropriate PR descriptions.

Updated the PR description. I'm sure that I already updated the description before make it ready for review 🤔

rafayet-monon commented:
> > Please add appropriate PR descriptions.
> 
> Updated the PR description. I'm sure that I already updated the description before make it ready for review 🤔

Don't know what happened but It was not shown to me 🤣

rafayet-monon commented on src/adapters/auth.ts:
```diff
@@ -20,6 +20,24 @@ class AuthAdapter {
   }
+
+  static forgotPassword = (email: string): Promise<AxiosResponse> => {
+    /* eslint-disable camelcase */
```
How about disabling this rule for the whole file?

gutakk replied:
Good catch. Fixed in 4bbbdbeccfec70ac04fdf24b76459cfff838dfd6

rafayet-monon commented on src/adapters/auth.ts:
```diff
@@ -20,6 +20,24 @@ class AuthAdapter {
+  static forgotPassword = (email: string): Promise<AxiosResponse> => {
+    /* eslint-disable camelcase */
+    const requestOptions = {
+      headers: {
```
Seems like every request is going to have this portion

```
headers: {
  'Content-Type': 'application/json'
}
```

Can we somehow add it `requestManager.ts` so that it isn't repeated?

gutakk replied:
I think we should not as the `requestManager.ts` is the generic function. We don't need this header when it comes to `GET` request or `POST` request without json body.

rafayet-monon replied:
If that's the case then how about making it DRY in some other way? As there are more API which will require json body. What do you think?

gutakk replied:
Fixed in f8142ea2b09043975ca0c0ed2a3b30a3046c4e81

rafayet-monon commented on src/components/BackButton/index.tsx:
```diff
@@ -0,0 +1,17 @@
+
+const BackButton = (): JSX.Element => {
+  const history = useHistory()
+  const goBackFunc = () => history.push('/')
```
Can we name this `goBack` instead of `goBackFunc`? 💭

gutakk replied:
Fixed in ccdac731d217e91054048b9af1a6bad9441c0c64

rafayet-monon commented on src/components/Background/index.tsx:
```diff
@@ -2,13 +2,17 @@ import React from 'react'
 type BackgroundProps = {
   backgroundImage: string
-  children: JSX.Element
+  children: JSX.Element | JSX.Element[]
```
why are we adding `JSX.Element[]`?

gutakk replied:
Because `ForgotPassword` page has more than one child element (back button and div in this case). Here is the error message when removing `JSX.Element[]`

> TypeScript error in /Users/gut/nimble/react-surveyor/src/screens/ForgotPassword/index.tsx(15,6):
This JSX tag's 'children' prop expects a single child of type 'Element', but multiple children were provided.  TS2746

rafayet-monon commented on src/screens/ForgotPassword/handler.ts:
```diff
@@ -0,0 +1,55 @@
+
+    if (!allFieldsFilled([target.email.value])) {
+      setError('Email is required')
+      setRequestSuccess(false)
```
`requestSuccess` is still `false` at this point so I guess it isn't needed here? 🤔

gutakk replied:
Because forgot password doesn't redirect to any page if the request success so when user successfully performs their request then `requestSuccess` will become to true and success alert will be displayed and then if user unsuccessfully perform their request then now success alert and error alert will displayed at the same time if I remove `setRequestSuccess(false)`

![Screen Shot 2564-06-22 at 12 30 46](https://user-images.githubusercontent.com/29707647/122868811-ad69b900-d355-11eb-84c8-da991896a3d4.png)

rafayet-monon commented on src/screens/ForgotPassword/handler.ts:
```diff
@@ -0,0 +1,55 @@
+        } else {
+          setError('Something went wrong. Please try again')
+        }
+        setRequestSuccess(false)
```
Same as the previous comment ⬆️

rafayet-monon commented on src/tests/components/Alert/index.test.tsx:
```diff
@@ -0,0 +1,41 @@
+
+describe('given Alert component is mounted', () => {
+  it('renders correct title', () => {
+    const element = render(<Alert title="someTitle" detail="someDetail" icon={alertIcon.bell} />)
```
How about destructuring?
```suggestion
    const { getByText } = render(<Alert title="someTitle" detail="someDetail" icon={alertIcon.bell} />)
```

gutakk replied:
Fixed in f788bf40fa566a2b845815d0da30e6a88704ff2d

Commit:
[Task-7] disable camel case lint for the whole file

Commit:
[Task-7] Update goBack function name

Commit:
[Task-7] Destruct const in the test

rafayet-monon commented on src/adapters/auth.ts:
```diff
@@ -20,6 +21,23 @@ class AuthAdapter {
+        user: {
+          email: email
+        },
+        client_id: process.env.REACT_APP_CLIENT_ID,
```
Let's create a function for this default values -

```
  static get DEFAULT_PAYLOAD() {
    return {
      client_id: process.env.REACT_APP_CLIENT_ID,
      client_secret: process.env.REACT_APP_CLIENT_SECRET
    };
  }
```

and use it here like `...this.DEFAULT_PAYLOAD`

gutakk replied:
Good idea. Fixed in 117eaa3ff205ed78d945b4ea1e86eba7f54a05b7
We cannot use `this` with `static`

> 'this' cannot be referenced in a static property initializer.ts(2334)

Commit:
[Task-7] Add default adapter payload

Commit:
[Task-7] Add adapter header

rafayet-monon commented on src/adapters/auth.ts:
```diff
@@ -1,25 +1,47 @@
 import requestManager from 'lib/requestManager'
 
 class AuthAdapter {
+  static DEFAULT_PAYLOAD = {
```
sorry for getting back to this again. `static` makes class level methods. As we don't need this outside the class how about making it a `const`. What do you think?

rafayet-monon reviewed (approved):
👍

Commit:
[Task-7] Add alert bell icon

Commit:
[Task-7] Update reset password to use AuthHeader
